File: 백준/Silver/9781. Knight Moves/Knight Moves.py
import sys
from collections import deque
input = sys.stdin.readline


def find(x):
    for i in range(n):
        for j in range(m):
            if board[i][j] == x:
                return i, j


def bfs(sx, sy, dx, dy):
    go = ((-2, -1), (-2, 1), (-1, -2), (-1, 2), (1, -2), (1, 2), (2, -1), (2, 1))

    q = deque()
    q.append((sx, sy, 0))

    visited = [[board[i][j] == '#' for j in range(m)] for i in range(n)]
    visited[sx][sy] = True

    while q:
        x, y, cnt = q.popleft()
        if (x, y) == (dx, dy):
            return cnt

        for a, b in go:
            nx, ny = x+a, y+b
            if nx < 0 or nx >= n or ny < 0 or ny >= m:
                continue
            if visited[nx][ny]:
                continue

            visited[nx][ny] = True
            q.append((nx, ny, cnt+1))
    return -1


n, m = map(int, input().split())
board = [input().rstrip() for _ in range(n)]

print(bfs(*find('K'), *find('X')))

File: 백준/Silver/9781. Knight Moves/Knight Moves.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Queue;


public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int[] nm = Arrays.stream(br.readLine().split(" "))
                .mapToInt(Integer::parseInt)
                .toArray();

        int n = nm[0];
        int m = nm[1];

        char[][] board = new char[n][];
        for (int i = 0; i < n; i++)
            board[i] = br.readLine().toCharArray();

        int[] start = find(board, n, m, 'K');
        int[] destination = find(board, n, m, 'X');

        System.out.println(bfs(board, n, m, start, destination));
    }

    private static int[] find(char[][] board, int n, int m, char want) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (board[i][j] == want)
                    return new int[]{i, j};
            }
        }
        return new int[]{-1, -1};
    }

    private static int bfs(char[][] board, int n, int m, int[] start, int[] destination) {
        int[] goX = {-2, -2, -1, -1, 1, 1, 2, 2};
        int[] goY = {-1, 1, -2, 2, -2, 2, -1, 1};

        int sx = start[0];
        int sy = start[1];
        int dx = destination[0];
        int dy = destination[1];

        Queue<int[]> q = new ArrayDeque<>();
        q.add(new int[]{sx, sy, 0});

        boolean[][] visited = new boolean[n][m];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (board[i][j] == '#')
                    visited[i][j] = true;
            }
        }
        visited[sx][sy] = true;

        while (!q.isEmpty()) {
            int[] first = q.poll();
            int x = first[0];
            int y = first[1];
            int cnt = first[2];

            if (x == dx && y == dy)
                return cnt;

            for (int goIdx = 0; goIdx < 8; goIdx++) {
                int nx = x + goX[goIdx];
                int ny = y + goY[goIdx];

                if (nx < 0 || nx >= n || ny < 0 || ny >= m)
                    continue;
                if (visited[nx][ny])
                    continue;

                visited[nx][ny] = true;
                q.add(new int[]{nx, ny, cnt+1});
            }
        }

        return -1;
    }
}